internal/pkg/types/v1alpha1/build.go

```diff
@@ -0,0 +1,216 @@
+package v1alpha1
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+	"golang.org/x/sys/unix"
+	"golang.org/x/xerrors"
+)
+
+func (p *Pkg) Build() error {
+	if err := environment(); err != nil {
+		return err
+	}
+
+	println("Build:", os.Getenv("BUILD"))
+	println("Host:", os.Getenv("HOST"))
+	println("Target:", os.Getenv("TARGET"))
+
+	for _, step := range p.Steps {
+		if _, err := os.Stat("/tmp"); err != nil {
+			if os.IsNotExist(err) {
+				// TODO(andrewrynhard): Remove this directory.
+				log.Println("Creating /tmp")
+				os.MkdirAll("/tmp", 0777)
+			}
+		}
+		dir, err := ioutil.TempDir("", "pkg")
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer os.RemoveAll(dir)
+
+		if err := os.Chdir(dir); err != nil {
+			log.Fatal(err)
+		}
+		if err := step.download(dir); err != nil {
+			log.Fatal(err)
+		}
+		if err := step.prepare(p.Shell); err != nil {
+			log.Fatal(err)
+		}
+		if err := step.build(p.Shell); err != nil {
+			log.Fatal(err)
+		}
+		if err := step.install(p.Shell); err != nil {
+			log.Fatal(err)
+		}
+		if err := step.test(p.Shell); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	return nil
+}
+
+func environment() error {
+	var result *multierror.Error
+
+	basic := func() {
+		os.Setenv("CFLAGS", "-g0 -Os")
+		os.Setenv("CXXFLAGS", "-g0 -Os")
+		os.Setenv("LDFLAGS", "-s")
+		os.Setenv("VENDOR", "talos")
+		os.Setenv("SYSROOT", "/talos")
+		os.Setenv("TOOLCHAIN", "/toolchain")
+		os.Setenv("PATH", fmt.Sprintf("/toolchain/bin:%s", os.Getenv("PATH")))
+	}
+
+	target := func() {
+		arch, ok := os.LookupEnv("TARGETPLATFORM")
+		if !ok {
+			result = multierror.Append(result, xerrors.New("TARGETPLATFORM is required"))
+			return
+		}
+
+		switch arch {
+		case "linux/amd64":
+			os.Setenv("ARCH", "x86_64")
+			os.Setenv("TARGET", "x86_64-talos-linux-musl")
+		case "linux/arm64":
+			os.Setenv("ARCH", "aarch64")
+			os.Setenv("TARGET", "aarch64-talos-linux-musl")
+		case "linux/arm/v7":
+			os.Setenv("ARCH", "armv7")
+			os.Setenv("TARGET", "armv7-talos-linux-musl")
+		default:
+			result = multierror.Append(result, xerrors.Errorf("unsupported build platform: %q", arch))
+			return
+		}
+	}
+
+	build := func() {
+		buf := &unix.Utsname{}
+		if err := unix.Uname(buf); err != nil {
+			result = multierror.Append(result, err)
+			return
+		}
+
+		b := bytes.Trim(buf.Machine[:], "\x00")
+		arch := string(b)
+		switch arch {
+		case "x86_64":
+			os.Setenv("BUILD", "x86_64-linux-musl")
+			os.Setenv("HOST", "x86_64-linux-musl")
+		case "aarch64":
+			os.Setenv("BUILD", "aarch64-linux-musl")
+			os.Setenv("HOST", "aarch64-linux-musl")
+		case "armv7l":
+			os.Setenv("BUILD", "armv7-linux-musl")
+			os.Setenv("HOST", "armv7-linux-musl")
+		default:
+			result = multierror.Append(result, xerrors.Errorf("unsupported build platform: %q", arch))
+			return
+		}
+	}
+
+	basic()
```

Comment:
this pattern with inline funcs seems odd/maybe even not right. I think it'd be clearer to just have these without the inline func as a series of switch statements.

Reply:
The inline funcs were just a way to help the readability. They really don't need to stand alone as individual funcs. The names of the func vars are hopefully descriptive and make it clear what they are for. Happy to implement another way, but looking for a way to make it clear that some are general, some are for build, and some are for target https://wiki.osdev.org/Target_Triplet.